sound150@claudiux/files/sound150@claudiux/applet.js

```diff
@@ -0,0 +1,1642 @@
+        let iconName = this._volumeToIcon(value);
+        if (this.app_icon == null) {
+            this.icon.icon_name = iconName;
+        }
+        this.setValue(value/this.applet.pcMaxVolume);
+
+        // send data to applet
+        this.emit("values-changed", iconName, percentage);
+    },
+
+    setValue: function(value) {
+        if (isNaN(value))
+            throw TypeError('The slider value must be a number');
+
+        this._value = Math.max(Math.min(value, 1), 0);
+        this._slider.queue_repaint();
+    },
+
+    _onScrollEvent: function (actor, event) {
+        let direction = event.get_scroll_direction();
+
+        if (direction == Clutter.ScrollDirection.DOWN) {
+            this._value = Math.max(0, this._value - VOLUME_ADJUSTMENT_STEP/this.applet.pcMaxVolume);
+        }
+        else if (direction == Clutter.ScrollDirection.UP) {
+            this._value = Math.min(1, this._value + VOLUME_ADJUSTMENT_STEP/this.applet.pcMaxVolume);
+        }
+
+        this._slider.queue_repaint();
+        this.emit('value-changed', this._value);
+    },
+
+    _moveHandle: function(absX, absY) {
+        let relX, relY, sliderX, sliderY;
+        [sliderX, sliderY] = this._slider.get_transformed_position();
+        relX = absX - sliderX;
+        relY = absY - sliderY;
+
+        let width = this._slider.width;
+        let handleRadius = this._slider.get_theme_node().get_length('-slider-handle-radius');
+
+        let newvalue;
+        if (relX < handleRadius)
+            newvalue = 0;
+        else if (relX > width - handleRadius)
+            newvalue = 1;
+        else
+            newvalue = (relX - handleRadius) / (width - 2 * handleRadius);
+        this._value = newvalue;
+        
+        this._slider.queue_repaint();
+        this.emit('value-changed', this._value);
+    },
+
+    get value() {
+        return this._value;
+    },
+
+    _onKeyPressEvent: function (actor, event) {
+        let key = event.get_key_symbol();
+        if (key == Clutter.KEY_Right || key == Clutter.KEY_Left) {
+            let delta = key == Clutter.KEY_Right ? VOLUME_ADJUSTMENT_STEP : -VOLUME_ADJUSTMENT_STEP;
+            this._value = Math.max(0, Math.min(this._value + delta/this.applet.pcMaxVolume, 1));
+            this._slider.queue_repaint();
+            this.emit('value-changed', this._value);
+            this.emit('drag-end');
+            return true;
+        }
+        return false;
+    },
+
+    _volumeToIcon: function(value){
+        if(value < 0.005)
+            return this.isMic? "microphone-sensitivity-none" : "audio-volume-muted";
+
+        let n = Math.floor(3 * value), icon;
+        if(n < 1)
+            icon = "low";
+        else if(n < 2)
+            icon = "medium";
+        else
+            icon = "high";
+
+        if (this.applet.percentMaxVolume > 100) {
+            let realValue = value * 100;
+            if (realValue <= 100) {
+                this.applet.actor.style_class = 'sound-normal';
```

Comment:
Saw this in ~/.xsession_errrors:

```
(cinnamon:1821): St-CRITICAL **: st_widget_get_theme_node called on the widget [0x55839f7ae440 StBoxLayout.sound-normal ("")] which is not in the stage.

(cinnamon:1821): St-CRITICAL **: st_widget_get_theme_node called on the widget [0x55839f7ae440 StBoxLayout.sound-normal ("")] which is not in the stage.

(cinnamon:1821): St-CRITICAL **: st_widget_get_theme_node called on the widget [0x55839f73f370 StBin:first-child] which is not in the stage.

(cinnamon:1821): St-CRITICAL **: st_widget_get_theme_node called on the widget [0x55839f7ae440 StBoxLayout.sound-normal ("")] which is not in the stage.

(cinnamon:1821): St-CRITICAL **: st_widget_get_theme_node called on the widget [0x55839f73f370 StBin:first-child] which is not in the stage.

(cinnamon:1821): St-CRITICAL **: st_widget_get_theme_node called on the widget [0x55839f86e940 StIcon.system-status-icon:last-child first-child] which is not in the stage.

(cinnamon:1821): St-CRITICAL **: st_widget_get_theme_node called on the widget [0x55839f7ae440 StBoxLayout.sound-normal ("")] which is not in the stage.

(cinnamon:1821): St-CRITICAL **: st_widget_get_theme_node called on the widget [0x55839f7ae440 StBoxLayout.sound-normal ("")] which is not in the stage.

(cinnamon:1821): St-CRITICAL **: st_widget_get_theme_node called on the widget [0x55839f73f370 StBin:first-child] which is not in the stage.

(cinnamon:1821): St-CRITICAL **: st_widget_get_theme_node called on the widget [0x55839f7ae440 StBoxLayout.sound-normal ("")] which is not in the stage.

(cinnamon:1821): St-CRITICAL **: st_widget_get_theme_node called on the widget [0x55839f73f370 StBin:first-child] which is not in the stage.

(cinnamon:1821): St-CRITICAL **: st_widget_get_theme_node called on the widget [0x55839f86e940 StIcon.system-status-icon:last-child first-child] which is not in the stage.
```

This usually means an St method manipulating style that relies on getting the theme node is called before the actor is staged (rendered) - it can't get style info yet at this point. Looking at where its being called, it looks like _update is getting called too early in either `on_settings_changed` or `connectWithStream`. You could try guarding the problem function with `if (!this.applet.actor.staged) return`.

Reply:
Hi,

It seems to be OK, now.

Many thanks to you !

Best regards,

claudiux